File: 프로그래머스/3/12971. 스티커 모으기（2）/스티커 모으기（2）.java
// import java.util.*;
class Solution {
    public int solution(int sticker[]) {
        int n = sticker.length;
        if (n == 1) return sticker[0];
        
        int[] dp1 = new int[n];
        dp1[0] = sticker[0];
        dp1[1] = sticker[0];
        
        int[] dp2 = new int[n];
        dp2[1] = sticker[1];
        
        for (int i = 2; i < n; i++) {
            if (i == n - 1) {
                dp1[i] = dp1[i - 1];
            } else {
                dp1[i] = Math.max(dp1[i - 2] + sticker[i], dp1[i - 1]);
            }
            
            dp2[i] = Math.max(dp2[i - 2] + sticker[i], dp2[i - 1]);
        }
        
        return Math.max(dp1[n - 1], dp2[n - 1]);
    }
}

File: 프로그래머스/3/12971. 스티커 모으기（2）/스티커 모으기（2）.js
function solution(sticker) {
    const len = sticker.length;
    const dp1 = new Array(sticker.length).fill(0);  // 첫번째 스티커 뗀 경우
    const dp2 = new Array(sticker.length).fill(0);  // 첫번째 스티커 떼지 않은 경우
    
    // 초기값 세팅
    dp1[0] = sticker[0];
    dp1[1] = sticker[0];
    dp2[1] = sticker[1];
    
    // 예외 처리
    if (len === 1) {
        return sticker[0];
    } else if (len === 2) {
        return Math.max(sticker[0], sticker[1]);
    }
    
    for (let i = 2; i < len; i++) {
        dp1[i] = Math.max(sticker[i] + dp1[i - 2], dp1[i - 1]);
        dp2[i] = Math.max(sticker[i] + dp2[i - 2], dp2[i - 1]);
        
        if (i === len - 1) dp1[i] = dp1[i - 1];
    }
    
    return Math.max(dp1[len - 1], dp2[len - 1]);
}